rpcs3/Emu/Cell/Modules/sceNpUtil.cpp

```diff
@@ -1,52 +1,155 @@
 #include "stdafx.h"
 #include "Emu/Cell/PPUModule.h"
+#include "Emu/Cell/lv2/sys_ppu_thread.h"
 #include "Emu/IdManager.h"
+#include "Emu/System.h"
 
 #include "sceNp.h"
 #include "sceNpUtil.h"
+#include "sysPrxForUser.h"
 
 LOG_CHANNEL(sceNpUtil);
 
-error_code sceNpUtilBandwidthTestInitStart(u32 prio, u64 stack)
+struct bandwidth_test final
+{
+	static const u32 id_base = 0xf0000000;
+	static const u32 id_step = 0x00000100;
+	static const u32 id_count = 1024;
+
+	atomic_t<u32> ppu_tid{};
+
+	atomic_t<bool> abort = false;
+	atomic_t<bool> shutdown = false;
+	atomic_t<bool> finished = false;
+
+	SceNpUtilBandwidthTestResult result{};
+
+	void exec(ppu_thread& ppu, u32 vid)
+	{
+		ppu_tid.release(ppu.id);
+
+		u32 fake_sleep_count = 0;
+
+		while (thread_ctrl::state() != thread_state::aborting)
+		{
+			if (abort || shutdown)
+			{
+				break;
+			}
+
+			// TODO: run proper bandwidth test, probably with cellHttp
+
+			thread_ctrl::wait_for(1000); // wait 1ms
+
+			if (fake_sleep_count++ > 100) // end fake bandwidth test after 100 ms
+			{
+				break;
+			}
+		}
+
+		result.upload_bps = 16000.0;
+		result.download_bps = 16000.0;
+		result.result = CELL_OK;
+
+		auto& util_manager = g_fxo->get<sce_np_util_manager>();
+		std::lock_guard lock(util_manager.mtx);
+		util_manager.status = SCE_NP_UTIL_BANDWIDTH_TEST_STATUS_FINISHED;
+		finished = true;
+	}
+};
+
+static void bandwidth_test_entry(ppu_thread& ppu, u32 vid)
+{
+	idm::get<bandwidth_test>(vid)->exec(ppu, vid);
+
+	ppu.state += cpu_flag::exit;
+}
+
+error_code sceNpUtilBandwidthTestInitStart(ppu_thread& ppu, u32 prio, u32 stack)
 {
 	sceNpUtil.todo("sceNpUtilBandwidthTestInitStart(prio=%d, stack=%d)", prio, stack);
```

Comment:
Since Shutdown and GetStatus are not writing a `todo`, this should also be either `notice` or `warning`

Reply:
the other two are pretty much implemented, while this one isn't

Reply:
ah. yeah. it kinda is

Reply:
I'd like to keep it on todo though. it's a telltale sign that the lib is used.
And since it's not really implemented, a higher visibility is probably better.